WordPress/Classes/Services/PostService.swift

```diff
@@ -0,0 +1,673 @@
+            self.managedObjectContext.performBlock({
+                if let postInContext = (try? self.managedObjectContext.existingObjectWithID(postObjectID)) as? AbstractPost {
+                    postInContext.remoteStatus = AbstractPostRemoteStatusFailed
+                    ContextManager.sharedInstance().saveContext(self.managedObjectContext)
+                }
+                failure(error)
+            })
+        }
+
+        if post.postID?.longLongValue > 0 {
+            remote?.updatePost(remotePost, success: successBlock, failure: failureBlock)
+        } else {
+            remote?.createPost(remotePost, success: successBlock, failure: failureBlock)
+        }
+    }
+
+    /**
+     Attempts to delete the specified post outright vs moving it to the
+     trash folder.
+
+     - parameters:
+         - post: The post or page to delete
+         - success: A success block
+         - failure: A failure block
+     */
+    func delete(post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+
+        let deleteBlock: SuccessBasic = {
+            if let postID = post.postID where postID.longLongValue > 0 {
+                let remotePost = self.getRemotePost(withPost: post)
+                if let remote = self.remote(for: post.blog) {
+                    remote.deletePost(remotePost, success: success, failure: failure)
+                }
+            }
+        }
+
+        if let original = post.original where post.isRevision() {
+            self.delete(original, success: deleteBlock, failure: failure)
+        } else {
+            deleteBlock()
+        }
+    }
+
+    /**
+     Moves the specified post into the trash bin. Does not delete
+     the post unless it was deleted on the server.
+
+     - parameters:
+         - post: The post or page to trash
+         - success: A success block
+         - failure: A failure block
+     */
+    func trash(post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+        if post.status == PostStatusTrash {
+            self.delete(post, success: success, failure: failure)
+            return
+        }
+
+        let trashBlock: SuccessBasic = {
+            if let status = post.status {
+                post.restorableStatus = status
+            }
+            if let postID = post.postID where postID.longLongValue <= 0  {
+                post.status = PostStatusTrash
+
+                success()
+
+                return
+            }
+
+            self.trashRemotePost(withPost: post, success: success, failure: failure)
+        }
+
+        if post.isRevision() {
+            self.trash(post, success: trashBlock, failure: failure)
+        } else {
+            trashBlock()
+        }
+    }
+
+    func trashRemotePost(withPost post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+        let postObjectID = post.objectID
+        let remotePost = self.getRemotePost(withPost: post)
+        if let remote = self.remote(for: post.blog) {
```

Comment:
Please, avoid stacking nested levels, since it's very very hard to read the code, otherwise.

`guard let` FTW!